lib/pure/ospaths.nim

```diff
@@ -157,64 +158,173 @@ const
-  ##   "usr/lib"
+  when doslikeFileSystem:
+    if path[0] in {DirSep, AltSep}:
+      return countWhile(path, a => a in {DirSep, AltSep})
+    if len(path) > 1 and path[0] in {'a'..'z', 'A'..'Z'} and path[1] == ':':
+      # eg: C:\\bar
+      return 2 + countWhile(path[2..^1], a => a in {DirSep, AltSep})
+  elif defined(macos):
+    # according to https://perldoc.perl.org/File/Spec/Mac.html `:a` is a relative path
+    if path[0] == ':':
+      result = 0
+    else:
+      result = countWhile(path, a => a != ':')
+  elif defined(RISCOS):
+    result = if path[0] == '$': 1 else: 0
+  elif defined(posix):
+    result = countWhile(path, a => a == '/')
+
+proc isAbsolute*(path: string): bool {.rtl, noSideEffect, extern: "nos$1".} =
+  ## Returns whether ``path`` is absolute.
   ##
-  ## If head is the empty string, tail is returned. If tail is the empty
-  ## string, head is returned with a trailing path separator. If tail starts
-  ## with a path separator it will be removed when concatenated to head. Other
-  ## path separators not located on boundaries won't be modified. More
-  ## examples on Unix:
+  ## On Windows, network paths are considered absolute too.
+  runnableExamples:
+    doAssert(not "".isAbsolute)
+    doAssert(not ".".isAbsolute)
+    doAssert(not "foo".isAbsolute)
+    when defined(posix):
+      doAssert "/".isAbsolute
+      doAssert(not "a/".isAbsolute)
+    when defined(Windows):
+      doAssert "C:\\foo".isAbsolute
+
+  return rootPrefixLength(path) > 0
+
+proc normalizePathEnd*(path:string, trailingSep = false): string =
+  ## Returns ``path`` with 0 or 1 trailing separator, depending on
+  ## ``trailingSep``, and taking care of special case of root paths.
+  runnableExamples:
+    when defined(posix):
+      doAssert "foo.bar//".normalizePathEnd == "foo.bar"
+      doAssert "bar//".normalizePathEnd(trailingSep = true) == "bar/"
+    when defined(Windows):
+      doAssert r"C:\foo\\".normalizePathEnd == r"C:\foo"
+      doAssert r"C:\foo".normalizePathEnd(trailingSep = true) == r"C:\foo\"
+      doAssert r"C:\".normalizePathEnd == r"C:\"
+  if path.len == 0: return ""
+
+  result = strip(path, leading = false, trailing = true, {DirSep, AltSep})
+  if trailingSep:
+    result = result & DirSep
+  else:
+    # still need to end with `DirSep` for root
+    if result.len == 0:
+      result = "" & DirSep
+    else:
+      when doslikeFileSystem:
+        if result[^1] == ':':
+          # C:\
+          result &= DirSep
+
+const absOverridesDefault = false
+
+proc joinPath*(head, tail: string, absOverrides = absOverridesDefault): string {.
+  noSideEffect, rtl, extern: "nos$1".} =
+  ## Concatenates paths ``head`` and ``tail``.
   ##
-  ## .. code-block:: nim
-  ##   assert joinPath("usr", "") == "usr/"
-  ##   assert joinPath("", "lib") == "lib"
-  ##   assert joinPath("", "/lib") == "/lib"
-  ##   assert joinPath("usr/", "/lib") == "usr/lib"
+  ## If ``tail`` is absolute and ``absOverrides`` is true, or ``head`` is empty,
```

Comment:
I'm not really sure about the use case for this.

Reply:
rationale is given in #8268 : most languages use the behavior absOverrides=true whereas before this PR nim was using the equivalent of absOverrides=false; so with this PR, 0 code is broken but user has option of using absOverrides=true